File: packages/store/src/conformance/compare.ts
// ─── Expectation comparison ─────────────────────────────────────────────────
// One comparator, used at generation time (does SQLite satisfy the expectation
// the memory reference produced?) and at replay time (does this backend satisfy
// the committed expectation?). Returns null on agreement, or a human-readable
// diff naming the JSON path that disagrees.
//
// Ordering, ids, counts and error messages are EXACT. Tolerance applies only to
// the fields a scenario names in `approxFields` — Mirk's parity contract is
// ordering, and a tolerant compare would hide the bug we most care about.
// `ignoreFields` drops a field from both sides entirely, for output that is
// real but not cross-backend contract.
//
// Throwing is decided here and nowhere else, symmetrically:
//   - a step whose expectation is `{throws}` PASSES only if the call raised and
//     the exception's message equals the pinned string exactly; if the call
//     returned normally the step FAILS;
//   - a step with any other expectation FAILS if the call raised.
// A step with no expectation at all is setup, and the caller enforces the same
// rule on it: setup that throws is a failure, not silence.

import {
  DEFAULT_TOL,
  isExpectIds,
  isExpectThrows,
  isExpectValue,
  isExpectValues,
  type Expect,
} from "./format.js";
import type { StepOutcome } from "./runner.js";

function show(value: unknown): string {
  return JSON.stringify(value) ?? String(value);
}

function typeName(value: unknown): string {
  if (value === null) return "null";
  if (Array.isArray(value)) return "array";
  return typeof value;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

/** Exact deep equality over JSON values, reporting the first divergent path.
 *  Integers and floats of equal value are equal (IEEE doubles, `===`). */
function deepDiff(actual: unknown, expected: unknown, path: string): string | null {
  if (typeName(actual) !== typeName(expected)) {
    return `at ${path}: expected ${typeName(expected)} ${show(expected)}, got ${typeName(actual)} ${show(actual)}`;
  }
  if (Array.isArray(expected) && Array.isArray(actual)) {
    if (actual.length !== expected.length) {
      return `at ${path}: expected length ${expected.length}, got ${actual.length} (${show(actual)})`;
    }
    for (let i = 0; i < expected.length; i++) {
      const diff = deepDiff(actual[i], expected[i], `${path}[${i}]`);
      if (diff) return diff;
    }
    return null;
  }
  if (isRecord(expected) && isRecord(actual)) {
    const keys = [...new Set([...Object.keys(expected), ...Object.keys(actual)])].sort();
    for (const key of keys) {
      if (!(key in expected)) {
        return `at ${path}.${key}: unexpected key with value ${show(actual[key])}`;
      }
      if (!(key in actual)) {
        return `at ${path}.${key}: missing key, expected ${show(expected[key])}`;
      }
      const diff = deepDiff(actual[key], expected[key], `${path}.${key}`);
      if (diff) return diff;
    }
    return null;
  }
  if (actual !== expected) {
    return `at ${path}: expected ${show(expected)}, got ${show(actual)}`;
  }
  return null;
}

/** Remove the ignored fields from a record. Applied to both sides, so an
 *  ignored field is not compared and not required to be present. */
export function stripIgnored(value: unknown, ignoreFields: readonly string[]): unknown {
  if (ignoreFields.length === 0 || !isRecord(value)) return value;
  const out: Record<string, unknown> = {};
  for (const [key, entry] of Object.entries(value)) {
    if (ignoreFields.includes(key)) continue;
    out[key] = entry;
  }
  return out;
}

function compareRecords(
  actual: unknown,
  expected: unknown[],
  approxFields: readonly string[],
  ignoreFields: readonly string[],
  tol: number,
): string | null {
  if (!Array.isArray(actual)) {
    return `at $: expected an array of ${expected.length} records, got ${typeName(actual)} ${show(actual)}`;
  }
  if (actual.length !== expected.length) {
    return `at $: expected length ${expected.length}, got ${actual.length} (${show(actual)})`;
  }
  for (let i = 0; i < expected.length; i++) {
    const expectedRow = stripIgnored(expected[i], ignoreFields);
    const actualRow = stripIgnored(actual[i], ignoreFields);
    if (!isRecord(expectedRow) || !isRecord(actualRow)) {
      const diff = deepDiff(actualRow, expectedRow, `$[${i}]`);
      if (diff) return diff;
      continue;
    }
    const keys = [...new Set([...Object.keys(expectedRow), ...Object.keys(actualRow)])].sort();
    for (const key of keys) {
      if (!(key in expectedRow)) {
        return `at $[${i}].${key}: unexpected key with value ${show(actualRow[key])}`;
      }
      if (!(key in actualRow)) {
        return `at $[${i}].${key}: missing key, expected ${show(expectedRow[key])}`;
      }
      if (approxFields.includes(key)) {
        const a = actualRow[key];
        const e = expectedRow[key];
        if (typeof a !== "number" || typeof e !== "number") {
          return `at $[${i}].${key}: approx field must be numeric, expected ${show(e)}, got ${show(a)}`;
        }
        if (Math.abs(a - e) > tol) {
          return `at $[${i}].${key}: expected ${e} within ${tol}, got ${a} (delta ${Math.abs(a - e)})`;
        }
        continue;
      }
      const diff = deepDiff(actualRow[key], expectedRow[key], `$[${i}].${key}`);
      if (diff) return diff;
    }
  }
  return null;
}

/** Check one step's outcome against its expectation. `null` means agreement. */
export function compareExpect(expect: Expect, actual: StepOutcome): string | null {
  if (isExpectThrows(expect)) {
    // Declared to throw. Returning normally is a failure: a validation rule
    // that stops firing must not pass as agreement.
    if (actual.ok) {
      return `expected a throw with message ${show(expect.throws)}, but the call returned ${show(actual.value)}`;
    }
    // Messages are contract. Exact string, not a code, class or prefix.
    if (actual.message !== expect.throws) {
      return `at $.message: expected ${show(expect.throws)}, got ${show(actual.message)}`;
    }
    return null;
  }

  // Not declared to throw. Raising is a failure, whatever the message says.
  if (!actual.ok) {
    return `expected a result, but the call threw ${show(actual.message)}`;
  }

  if (isExpectValue(expect)) {
    return deepDiff(actual.value, expect.value, "$");
  }
  if (isExpectValues(expect)) {
    return compareRecords(
      actual.value,
      expect.values,
      expect.approxFields ?? [],
      expect.ignoreFields ?? [],
      expect.tol ?? DEFAULT_TOL,
    );
  }
  if (isExpectIds(expect)) {
    if (!Array.isArray(actual.value)) {
      return `at $: expected an array of records, got ${typeName(actual.value)} ${show(actual.value)}`;
    }
    const ids = actual.value.map((row) => (isRecord(row) ? row.id : undefined));
    return deepDiff(ids, expect.ids, "$.ids");
  }

  return `unrecognized expect form ${show(expect)}`;
}

File: python/store/src/mirk/store/conformance/compare.py
"""Compare a step result against a corpus ``expect`` clause.

Every function returns ``None`` on a match, or a human-readable diff naming the
JSON path that disagreed.

Forms: ``value`` (exact), ``values`` with optional ``approxFields``/``tol``
(compared within tolerance) and optional ``ignoreFields`` (dropped from both
sides before comparison), ``ids`` (ordered result ids), and ``throws`` (exact
message).
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, cast

if TYPE_CHECKING:
    from .runner import StepOutcome

__all__ = ["compare_expect", "deep_equal"]

_DEFAULT_TOL = 1e-6


def _kind(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "bool"
    if isinstance(value, int | float):
        return "number"
    if isinstance(value, str):
        return "string"
    if isinstance(value, list):
        return "array"
    if isinstance(value, dict):
        return "object"
    return type(value).__name__


def _as_list(value: Any) -> list[Any]:
    return cast(list[Any], value)


def _as_object(value: Any) -> dict[str, Any]:
    return cast(dict[str, Any], value)


def _is_number(value: Any) -> bool:
    return isinstance(value, int | float) and not isinstance(value, bool)


def _field_names(value: Any) -> list[str]:
    return [str(field) for field in _as_list(value)] if isinstance(value, list) else []


def deep_equal(actual: Any, expected: Any, path: str = "$") -> str | None:
    """Exact deep equality. Bools are not numbers; ints and floats compare by value."""
    if isinstance(expected, bool) or isinstance(actual, bool):
        if actual is expected:
            return None
        return f"{path}: expected {expected!r}, got {actual!r}"
    if expected is None or actual is None:
        if actual is None and expected is None:
            return None
        return f"{path}: expected {expected!r}, got {actual!r}"
    if _is_number(expected) and _is_number(actual):
        return None if actual == expected else f"{path}: expected {expected!r}, got {actual!r}"
    if _kind(actual) != _kind(expected):
        return f"{path}: expected {_kind(expected)} {expected!r}, got {_kind(actual)} {actual!r}"
    if isinstance(expected, str):
        return None if actual == expected else f"{path}: expected {expected!r}, got {actual!r}"
    if isinstance(expected, list):
        actual_items = _as_list(actual)
        expected_items = _as_list(expected)
        if len(actual_items) != len(expected_items):
            return f"{path}: expected {len(expected_items)} items, got {len(actual_items)}"
        for index in range(len(expected_items)):
            diff = deep_equal(actual_items[index], expected_items[index], f"{path}[{index}]")
            if diff is not None:
                return diff
        return None
    if isinstance(expected, dict):
        actual_obj = _as_object(actual)
        expected_obj = _as_object(expected)
        for key in sorted(set(actual_obj) | set(expected_obj)):
            if key not in actual_obj:
                return f"{path}.{key}: missing, expected {expected_obj[key]!r}"
            if key not in expected_obj:
                return f"{path}.{key}: unexpected {actual_obj[key]!r}"
            diff = deep_equal(actual_obj[key], expected_obj[key], f"{path}.{key}")
            if diff is not None:
                return diff
        return None
    return None if actual == expected else f"{path}: expected {expected!r}, got {actual!r}"


def _compare_approx(
    actual: Any,
    expected: Any,
    approx_fields: list[str],
    ignore_fields: list[str],
    tol: float,
    path: str,
) -> str | None:
    if not isinstance(actual, list):
        return f"{path}: expected an array, got {_kind(actual)}"
    actual_rows = _as_list(actual)
    expected_rows = _as_list(expected)
    if len(actual_rows) != len(expected_rows):
        return f"{path}: expected {len(expected_rows)} rows, got {len(actual_rows)}"
    for index in range(len(expected_rows)):
        row: Any = actual_rows[index]
        want: Any = expected_rows[index]
        row_path = f"{path}[{index}]"
        if not isinstance(row, dict) or not isinstance(want, dict):
            diff = deep_equal(row, want, row_path)
            if diff is not None:
                return diff
            continue
        dropped = set(ignore_fields)
        row_obj = {k: v for k, v in _as_object(row).items() if k not in dropped}
        want_obj = {k: v for k, v in _as_object(want).items() if k not in dropped}
        # Key sets are compared in both directions, so an approx field present in
        # the actual row but absent from the expected row is a failure, not a
        # field to drop.
        for field in sorted(set(row_obj) | set(want_obj)):
            if field not in want_obj:
                return f"{row_path}.{field}: unexpected {row_obj[field]!r}"
            if field not in row_obj:
                return f"{row_path}.{field}: missing, expected {want_obj[field]!r}"
            if field in approx_fields:
                got: Any = row_obj[field]
                target: Any = want_obj[field]
                if not _is_number(got) or not _is_number(target):
                    return f"{row_path}.{field}: expected a number, got {got!r}"
                if abs(float(got) - float(target)) > tol:
                    return f"{row_path}.{field}: {got!r} is not within {tol} of {target!r}"
                continue
            diff = deep_equal(row_obj[field], want_obj[field], f"{row_path}.{field}")
            if diff is not None:
                return diff
    return None


def compare_expect(outcome: StepOutcome, expect: dict[str, Any]) -> str | None:
    """Check one step outcome against its ``expect`` clause.

    Only a :class:`~mirk.store.conformance.runner.StepOutcome` is accepted: the
    raise/return distinction is carried by the outcome, never guessed from the
    shape of a returned value.
    """
    from .runner import StepOutcome as _StepOutcome  # local: runner imports this module

    if not isinstance(outcome, _StepOutcome):  # pyright: ignore[reportUnnecessaryIsInstance]
        raise TypeError(f"compare_expect needs a StepOutcome, got {type(outcome).__name__}")

    if "throws" in expect:
        if outcome.ok:
            return f"expected a raise with {expect['throws']!r}, got {outcome.value!r}"
        if outcome.message != expect["throws"]:
            return f"$.message: expected {expect['throws']!r}, got {outcome.message!r}"
        return None

    if not outcome.ok:
        return f"unexpected raise: {outcome.message!r}"

    actual: Any = outcome.value

    if "value" in expect:
        return deep_equal(actual, expect["value"])

    if "values" in expect:
        approx = _field_names(expect.get("approxFields"))
        ignore = _field_names(expect.get("ignoreFields"))
        tol_raw: Any = expect.get("tol")
        tol = float(tol_raw) if _is_number(tol_raw) else _DEFAULT_TOL
        return _compare_approx(actual, expect["values"], approx, ignore, tol, "$")

    if "ids" in expect:
        if not isinstance(actual, list):
            return f"$: expected an array, got {_kind(actual)}"
        ids: list[Any] = []
        for row in _as_list(actual):
            ids.append(_as_object(row).get("id") if isinstance(row, dict) else row)
        return deep_equal(ids, expect["ids"], "$.ids")

    return f"unsupported expect form: {sorted(expect)}"
